### PR TITLE
runtime: Fix /var/lib/vc/sbs/${sid} dir residual

### PR DESCRIPTION
**Pull request by keloyang**
Fixes: #2921

runtime call fetchSandbox-->loadSandboxConfigFromOldStore-->store.NewVCSandboxStore--> ... f.initialize -->  os.MkdirAll(f.path, DirMode) 
and at last create /var/lib/vc/sbs/${sid}, but don't delete it before delete sandbox.

Signed-off-by: Shukui Yang <keloyangsk@gmail.com>

### CODE REVIEW

**Review by fidencio — COMMENTED** [review]
@keloyang, I'm not sure if I understand this fix.
Would you mind to elaborate why it's needed? Please, also add the details to the commit message.

**Comment by keloyang** [comment]
@fidencio I have updated the issue #2921, PTAL, thanks.

**Comment by jodh-intel** [comment]
@keloyang - Thanks for raising! A couple of things:

- Is there any way we can add a unit test here to ensure the correct behaviour?
- Could you also port this to Kata 2.0 (https://github.com/kata-containers/kata-containers/tree/2.0-dev/src/runtime)?

@bergwolf, @sameo - could you tal at this change please?

**Comment by fidencio** [comment]
Let me give it a  try on my environment here, thanks for the update @keloyang.

**Review by amshinde — APPROVED** [review]

**Comment by amshinde** [comment]
@keloyang Thanks for the fix!  LGTM. Could you please add add a unit/integration test verifying this behaviour?

**Comment by keloyang** [comment]
@amshinde @jodh-intel  unit test case added, ptal, thanks.

**Review by fidencio — COMMENTED** [review]
I'd like to ask @amshinde, @jodh-intel, to not merge this one yet!

I want to try to reproduce this one and will most likely find some time this evening to do so.
There are a few things puzzling me here, which I'm commenting inline.

File: virtcontainers/sandbox.go
Comment (fidencio) [inline virtcontainers/sandbox.go:840]: Do we really have to create a new vcStore or should we just use s.store?

File: virtcontainers/sandbox.go
Comment (fidencio) [inline virtcontainers/sandbox.go:842]: We definitely should check for errors here, at least to log them.

File: virtcontainers/sandbox_test.go
Comment (jodh-intel) [inline virtcontainers/sandbox_test.go:1559]: I think it would be clearer to write this as separate tests:

```go
err := os.Stat(runtimSidPath)
assert.Error(err)
assert.True(os.IsNotExist(err))
```

**Review by jodh-intel — COMMENTED** [review]
Thanks @keloyang - just one comment.

**Comment by jodh-intel** [comment]
@fidencio - agreed - added `dnm` for now.

**Comment by keloyang** [comment]
> Do we really have to create a new vcStore or should we just use s.store?

runtime set s.store only when  useOldStore(ctx) is true in https://github.com/kata-containers/runtime/blob/master/virtcontainers/sandbox.go#L573, this need loadSandboxConfigFromOldStore return nil, but it return error if old store don't exist. this means everytime when fetchSandbox is called ,it will create a vcStore, but s.store is nil. so we can't use s.store  simplely. 
loadSandboxConfigFromOldStore is for backford,and old store has been deprecated, so it's better to use loadSandboxConfig before loadSandboxConfigFromOldStore, so I have new change for this, ptal.

File: virtcontainers/sandbox.go
Comment (fidencio) [inline virtcontainers/sandbox.go:731]: @keloyang, what was the reason of inverting the logic here?
I think we still should try to load from the old store at first.

File: virtcontainers/sandbox.go
Comment (fidencio) [inline virtcontainers/sandbox.go:840]: @keloyang what's the reason of only doing this for an old store?

**Review by fidencio — COMMENTED** [review]
@keloyang,

I've made some comments inline, but I must admit I'm failing to reproduce the issues.

What's exactly the version of the runtime you're using that you can reproduce it?

File: virtcontainers/sandbox.go
Comment (keloyang) [reply virtcontainers/sandbox.go:731]: The root cause of the residue is loadSandboxConfigFromOldStore is called but old store file is not exist,  and if load sandbox from new store successfull, there is no need to call loadSandboxConfigFromOldStore.

File: virtcontainers/sandbox.go
Comment (keloyang) [reply virtcontainers/sandbox.go:840]: useOldStore  return true only loadSandboxConfigFromOldStore return successfully,  so I add useOldStore here, but it's not necessary.

**Comment by keloyang** [comment]
@fidencio  please show me how do you try to reproduce,  I use the master branch, and it's easy to reproduce.
```
[root@centos1 ~]# ls  /var/lib/vc/sbs/                                               
[root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/

15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa
[root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/

15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa  cd86011f6c0b779a2e3f40b49fa06463757ac1413ba70007e14462b7d0782e24
[root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/

15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa  93fb2abfbe151c8f12e3b2b7f24321479b2425af80acb93f2f4c2263fb679621  cd86011f6c0b779a2e3f40b49fa06463757ac1413ba70007e14462b7d0782e24
```

File: virtcontainers/sandbox.go
Comment (fidencio) [reply virtcontainers/sandbox.go:731]: @amshinde, @egernst,

Let me ask some more experienced people for some help.

I think we still should try to load the old store first.

In that very same code path we call `createSandbox()`, which calls `newSandbox()` and there depending on whether we use the old store or not we'd call `store.NewCVSandboxStore()`.

My question here is, why we only call that for one case?

**Comment by keloyang** [comment]
ping @jodh-intel @amshinde @fidencio

**Comment by jodh-intel** [comment]
@keloyang - please check the CI's at the end of your PR. Travis is failing on this one too:

```
virtcontainers/sandbox.go:840: File is not `gofmt`-ed with `-s` (gofmt)
```

File: virtcontainers/sandbox.go
Comment (fidencio) [reply virtcontainers/sandbox.go:731]: @jodh-intel, maybe you can help me with this one? ^^

**Comment by fidencio** [comment]
> @fidencio please show me how do you try to reproduce, I use the master branch, and it's easy to reproduce.
> 
> ```
> [root@centos1 ~]# ls  /var/lib/vc/sbs/                                               
> [root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/
> 
> 15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa
> [root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/
> 
> 15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa  cd86011f6c0b779a2e3f40b49fa06463757ac1413ba70007e14462b7d0782e24
> [root@centos1 ~]# docker run --rm -ti --runtime untrusted-runtime 018c9d7b792b echo; ls  /var/lib/vc/sbs/
> 
> 15760f7e07c98e4f4ed9d7aba4076de14475dd71a61be9919b24037c810b37aa  93fb2abfbe151c8f12e3b2b7f24321479b2425af80acb93f2f4c2263fb679621  cd86011f6c0b779a2e3f40b49fa06463757ac1413ba70007e14462b7d0782e24
> ```

I was able to reproduce the issue and I raised a comment on a part that I don't have a clear understanding. Let's wait till the other developers jump into that.

**Comment by keloyang** [comment]
> virtcontainers/sandbox.go

updated and sorry for not gofmt, thanks @jodh-intel

**Review by jodh-intel — APPROVED** [review]
Thanks @keloyang.

lgtm

**Comment by jodh-intel** [comment]
@keloyang - I've restarted the failing metrics CI to see if this PR now passes. Please can you port this to 2.0 (https://github.com/kata-containers/kata-containers), or add a link to the 2.0 PR here if already done?

**Comment by amshinde** [comment]
I had tested a previous version of this PR, will like to take another look before this is merged.

File: virtcontainers/sandbox.go
Comment (fidencio) [reply virtcontainers/sandbox.go:731]: @amshinde, @jodh-intel, @egernst,
ping about this question.

Comment (amshinde) [reply virtcontainers/sandbox.go:731]: @fidencio This patch makes sense to me. We have kept the old store code around for backward compatibility. But by checking for the old store and trying to initialize one and checking for failure while using the old store, we were inadvertently creating the directory structure under `/var/lib/vc/sbs/{sandbox_id}` and leaving that around after the sandbox is deleted. 
It makes sense to me to check for the new store first instead. I tested this patch and works as expected.

**Comment by amshinde** [comment]
lgtm @fidencio. I think this one is ready to be merged.

**Comment by keloyang** [comment]
I think there is no need to do porting for 2.0-dev, because there is no code to call loadSandboxConfigFromOldStore in fetchSandbox, see https://github.com/kata-containers/kata-containers/blob/2.0-dev/src/runtime/virtcontainers/sandbox.go#L650.
ping @jodh-intel @amshinde @fidencio

**Comment by jodh-intel** [comment]
@keloyang - Regarding, https://github.com/kata-containers/runtime/pull/2922#issuecomment-707636127, I think you are correct, so I'm removing the `needs-forward-port` label...

**Pull request merged**
